cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        .build();
+    }
+
+    /**
+     * Cancel the currently scheduled stream size polling task, and reschedule one for later.
+     */
+    private void cancelPollingAndScheduleNext() {
+      // This method might be called from the call to #received defined in the below Runnable - in which case
+      // this scheduledPolling would in fact be active. Hence we don't want to interrupt the active task
+      if (scheduledPolling != null) {
+        scheduledPolling.cancel(false);
+      }
+
+      // Regardless of whether cancelling was successful, we still want to schedule the next polling
+      scheduledPolling = streamPollingExecutor.schedule(createPollingRunnable(), pollingDelay, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * @return a runnable that uses the {@link StreamAdmin} to poll the stream size, and creates a fake notification
+     *         with that size, so that this information can be treated as if it came from a real notification.
+     */
+    private Runnable createPollingRunnable() {
+      return new Runnable() {
+        @Override
+        public void run() {
+          // We only perform polling if at least one scheduleTask is active
+          if (activeTasks.get() > 0) {
+            try {
+              StreamSize streamSize = getStreamEventsSize();
+
+              // We don't need a notification context here
+              received(new StreamSizeNotification(streamSize.getTimestamp(), streamSize.getSize()), null);
+            } catch (IOException e) {
+              LOG.debug("Ignoring stream events size polling after error", e);
+            }
+          }
+        }
+      };
+    }
+
+    /**
+     * @return size of events ingested by the stream so far, queried using the metric system
+     */
+    private StreamSize getStreamEventsSize() throws IOException {
+      MetricDataQuery metricDataQuery = new MetricDataQuery(
+        0L, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()),
+        Integer.MAX_VALUE, "system.collect.bytes",
+        MetricType.COUNTER,
+        ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, streamId.getNamespaceId(),
+                        Constants.Metrics.Tag.STREAM, streamId.getName()),
+        ImmutableList.<String>of()
+      );
+
+      try {
+        Collection<MetricTimeSeries> metrics = metricStore.query(metricDataQuery);
+        if (metrics == null || metrics.isEmpty()) {
+          // Data is not yet available, which means no data has been ingested by the stream yet
+          return new StreamSize(0L, System.currentTimeMillis());
+        }
+
+        MetricTimeSeries metric = metrics.iterator().next();
+        List<TimeValue> timeValues = metric.getTimeValues();
+        if (timeValues == null || timeValues.size() != 1) {
+          throw new IOException("Should collect exactly one time value");
+        }
+        TimeValue timeValue = timeValues.get(0);
+        // The metric store gives us 0 as the timestamp, hence we cannot use it here
```

Comment:
I still think this warrants a Jira. It is a functionality that we want, and the metrics system should be requirement driven. Metrics team can always comment on the Jira and explain why that request makes no sense.